File: src/MyServiceBus.AmazonSqs/AmazonSqsEndpointAddress.cs
namespace MyServiceBus;

internal enum AmazonSqsEntityKind
{
    Queue,
    Topic
}

internal readonly record struct AmazonSqsEndpointAddress(string EntityName, AmazonSqsEntityKind Kind)
{
    public static AmazonSqsEndpointAddress Parse(Uri address)
    {
        ArgumentNullException.ThrowIfNull(address);
        if (!address.Scheme.Equals("amazonsqs", StringComparison.OrdinalIgnoreCase) &&
            !address.Scheme.Equals("queue", StringComparison.OrdinalIgnoreCase) &&
            !address.Scheme.Equals("topic", StringComparison.OrdinalIgnoreCase))
            throw new ArgumentException($"Unsupported Amazon SQS address scheme '{address.Scheme}'.", nameof(address));

        var logicalScheme = address.Scheme.Equals("queue", StringComparison.OrdinalIgnoreCase) ||
                            address.Scheme.Equals("topic", StringComparison.OrdinalIgnoreCase);
        var name = logicalScheme
            ? address.OriginalString[(address.Scheme.Length + 1)..].Split('?', 2)[0].TrimStart('/')
            : address.Segments.LastOrDefault()?.Trim('/');
        AmazonSqsEntityName.Validate(name ?? string.Empty);

        var query = System.Web.HttpUtility.ParseQueryString(address.Query);
        var kind = address.Scheme.Equals("topic", StringComparison.OrdinalIgnoreCase) ||
                   string.Equals(query["type"], "topic", StringComparison.OrdinalIgnoreCase)
            ? AmazonSqsEntityKind.Topic
            : AmazonSqsEntityKind.Queue;
        return new AmazonSqsEndpointAddress(name!, kind);
    }
}

File: src/Java/myservicebus-amazon-sqs/src/main/java/com/myservicebus/amazon/sqs/AmazonSqsEndpointAddress.java
package com.myservicebus.amazon.sqs;

import java.net.URI;

record AmazonSqsEndpointAddress(String entityName, EntityKind kind) {
    enum EntityKind { QUEUE, TOPIC }

    static AmazonSqsEndpointAddress parse(URI address) {
        String scheme = address.getScheme().toLowerCase(java.util.Locale.ROOT);
        if (!scheme.equals("amazonsqs") && !scheme.equals("queue") && !scheme.equals("topic")) {
            throw new IllegalArgumentException("Unsupported Amazon SQS address scheme: " + scheme);
        }
        String path = address.isOpaque() ? address.getSchemeSpecificPart().split("\\?", 2)[0] : address.getPath();
        String name = path.substring(path.lastIndexOf('/') + 1);
        AmazonSqsEntityNames.validate(name);
        boolean topic = scheme.equals("topic") || (address.getQuery() != null &&
                java.util.Arrays.stream(address.getQuery().split("&"))
                        .anyMatch(value -> value.equalsIgnoreCase("type=topic")));
        return new AmazonSqsEndpointAddress(name, topic ? EntityKind.TOPIC : EntityKind.QUEUE);
    }
}
